pkg/ha/ha_service.go

```diff
@@ -0,0 +1,191 @@
+package ha
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/jackc/pgconn"
+	"github.com/timescale/promscale/pkg/ha/state"
+
+	"golang.org/x/sync/semaphore"
+
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgxconn"
+)
+
+const (
+	backOffDurationOnLeaderChange = 10 * time.Second
+	haSyncerTimeInterval          = 15 * time.Second
+)
+
+type Service struct {
+	state             *sync.Map
+	lockClient        haLockClient
+	leaseTimeout      time.Duration
+	leaseRefresh      time.Duration
+	_leaderChangeLock *semaphore.Weighted
+	_syncTicker       *time.Ticker
+}
+
+func NewHAService(dbClient pgxconn.PgxConn) (*Service, error) {
+	lockClient := newHaLockClient(dbClient)
+	timeout, refresh, err := lockClient.readLeaseSettings(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("could not create HA state: %#v", err)
+	}
+
+	service := &Service{
+		state:             &sync.Map{},
+		lockClient:        lockClient,
+		leaseTimeout:      timeout,
+		leaseRefresh:      refresh,
+		_leaderChangeLock: semaphore.NewWeighted(1),
+		_syncTicker:       time.NewTicker(haSyncerTimeInterval),
+	}
+	go service.haStateSyncer()
+	return service, nil
+}
+
+func (h *Service) haStateSyncer() {
+	for range h._syncTicker.C {
+		h.state.Range(func(c, s interface{}) bool {
+			cluster := fmt.Sprint(c)
+			state := castToState(s)
+			stateView := state.Clone()
+			err := h.syncLockStateFromDB(stateView.LeaseStart, stateView.MaxTimeSeenLeader, cluster, stateView.Leader)
+			if err != nil {
+				log.Error("failed to validate cluster %s state %v", cluster, err)
+			}
+			if stateView.MaxTimeSeen.After(stateView.LeaseUntil) {
+				go h.tryChangeLeader(cluster, stateView.Leader)
+			}
+			return true
+		})
+	}
+}
+
+// checkInsert verifies the samples are from prom leader & in an expected time range.
+// 	returns an boolean signifying whether to allow (true),
+//		or deny (false). The second returned argument is the
+//      minimum timestamp of the accepted samples.
+//		An error is returned if the lock state could not be
+//		checked against the db
+func (h *Service) CheckInsert(minT, maxT time.Time, clusterName, replicaName string) (bool, time.Time, error) {
+	s, ok := h.state.Load(clusterName)
+	if !ok {
+		s, _ = h.state.LoadOrStore(clusterName, &state.State{})
+		err := h.syncLockStateFromDB(minT, maxT, clusterName, replicaName)
+		if err != nil {
+			return false, time.Time{}, err
+		}
+	}
+
+	state := castToState(s)
+	stateView := state.Clone()
+
+	defer state.UpdateMaxSeenTime(replicaName, maxT)
+	if replicaName != stateView.Leader {
+		return false, time.Time{}, nil
+	}
+
+	acceptedMinT := minT
+	if minT.Before(stateView.LeaseStart) {
+		acceptedMinT = stateView.LeaseStart
+	}
+
+	if maxT.After(stateView.LeaseUntil) {
+		err := h.syncLockStateFromDB(minT, maxT, clusterName, replicaName)
```

Comment:
hmm what happens if there was a leader change instead of a successful update here? In that case there won't be an error but the current return would be wrong. I think we have to re-check who is the leader after the call to  h.syncLockStateFromDB.